rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md

```diff
@@ -0,0 +1,542 @@
+        shuffle=hp.Boolean("shuffle"),
+        # Tune whether to use sample_weights.
+        sample_weight=sample_weight if hp.Boolean("sample_weight") else None,
+        # The provided callbacks list contains checkpointing and tensorboard.
+        callbacks=callbacks)
+
+tuner = kt.RandomSearch(
+  hypermodel=MyHyperModel(),
+  objective=kt.Objective('val_mae', 'min'),
+  directory='dir',
+  max_trials=3,
+  executions_per_trial=2,
+  overwrite=True,
+  directory="my_dir",
+  project_name="helloworld",
+)
+
+# Any arg passed to `search()` would be passed to `fit()`.
+tuner.search()
+
+# Exporting the best models.
+models = tuner.get_best_models(num_models=2)
+
+# Retraining the model with the second best hyperparameters.
+second_best_hp = tuner.get_best_hyperparameters(num_models=2)[1]
+hypermodel = MyHyperModel()
+model = hypermodel.build(second_best_hp)
+hypermodel.fit(model, hp=second_best_hp, callbacks=[], dataset=new_dataset)
+```
+
+There are three details in this workflow that are worth attention.
+
+* Similar to `Tuner.run_trial()`, the return value of the fit function supports
+  all different formats.
+
+* The user can use built-in callbacks just like in `Tuner.run_trial()`.
+
+* `build()` and `fit()` can share hyperparameters. In this example,
+  `num_features` is shared between the two functions. In `fit()`, we can use
+  `hp.get()` to obtain the value of a hyperparameter used in `build()`.
+
+* We can easily retrain the model with any hyperparameter value set with
+  `hypermodel.build()` and `hypermodel.fit()`.
+
+With these proposed workflows, the user now has the maximum flexibility. Any
+step in an end-to-end machine learning workflow can be tuned. Moreover, the
+changes needed to tune existing Keras code is minimized.
+
+Here we present HyperModel code examples of three important use cases:
+
+* Text tokenization.
+
+* Custom training loop.
+
+* Fine tuning with pretrained weights.
+
+#### Text tokenization
+
+```py
+class MyHyperModel(kt.HyperModel):
+  def build(self, hp):
+    inputs = keras.Input(shape=(10,))
+    outputs = layers.Embedding(
+        # max_token is a hyperparameter also used in text vectorization.
+        input_dim=hp.Int("max_tokens", 100, 500, step=100),
+        output_dim=64)(inputs)
+    outputs = layers.LSTM(hp.Int("units", 32, 128, step=32))(outputs)
+    outputs = layers.Dense(1, activation='sigmoid')(outputs)
+    model = keras.Model(inputs, outputs)
+    model.compile(loss='mse')
+    return model
+  
+  def fit(self, hp, model, x, y, validation_data, callbacks, **kwargs):
+    # Create and adapt the text vectorizer.
+    text_vectorizer = layers.TextVectorization(
+        # The max_tokens is a hyperparameter created in build().
+        max_tokens=hp.get("max_tokens"),
+        output_mode="int",
+        output_sequence_length=10)
+    text_vectorizer.adapt(x)
+  
+    # Save the text vectorization model.
+    keras.Sequential([text_vectorizer]).save(
```

Comment:
You'd just save the vocab file as a text file, not a Sequential.

Reply:
@mattdangerw Is there any code examples for saving, loading, and truncating the vocab instead of the layer?

Reply:
Hmm, you could save a vocabulary for a TextVectorization to a file like this.

```
vocab = text_vectorizer.get_vocabulary(include_special_tokens=False)
with open('vocab.txt', 'w') as file:
  for term in vocab:
    file.write(term + '\n')
```

But the issue is we don't really support loading vocabulary file only up to a truncated line. I don't believe the underlying op supports this either though would need to check.

Things that would work today.
 1) save file, truncate file by hand after saving (maybe awkward)
 2) truncate vocabulary by creating new TextVectorization layers
 3) truncate only output of layer

```
adapted_tv = layers.TextVectorization()
adapted_tv.adapt(data)
# option 2, truncate by creating new layer
vocab = adapted_tv.get_vocabulary()
truncated_tv = layers.TextVectorization(vocabulary=vocab[:max_tokens])

# option 3, truncate output from layer
def apply_tv(input):
  output = adapted_tv(input)
  # Truncate layer output to smaller vocab size. 1 is the oov token.
  return tf.where(output >= max_tokens, tf.ones_like(output), output)
```

Would any of those make sense? Seems like truncating layer output might be easiest.